src/main/java/org/apache/cassandra/sidecar/cluster/SidecarLoadBalancingPolicy.java

```diff
@@ -0,0 +1,217 @@
+ */
+class SidecarLoadBalancingPolicy implements LoadBalancingPolicy
+{
+    public static final int MIN_ADDITIONAL_CONNECTIONS = 2;
+    private static final Logger LOGGER = LoggerFactory.getLogger(SidecarLoadBalancingPolicy.class);
+    private final Set<Host> selectedHosts = new HashSet<>();
+    private final Set<InetSocketAddress> localHostAddresses;
+    private final LoadBalancingPolicy childPolicy;
+    private final int totalRequestedConnections;
+    private final Random secureRandom = new SecureRandom();
+    private final HashSet<Host> allHosts = new HashSet<>();
+    private Cluster cluster;
+
+    public SidecarLoadBalancingPolicy(List<InetSocketAddress> localHostAddresses,
+                                      String localDc,
+                                      int numAdditionalConnections)
+    {
+        this.childPolicy = createChildPolicy(localDc);
+        this.localHostAddresses = new HashSet<>(localHostAddresses);
+        if (numAdditionalConnections < MIN_ADDITIONAL_CONNECTIONS)
+        {
+            LOGGER.warn("Additional instances requested was {}, which is less than the minimum of {}. Using {}.",
+                        numAdditionalConnections, MIN_ADDITIONAL_CONNECTIONS, MIN_ADDITIONAL_CONNECTIONS);
+            numAdditionalConnections = MIN_ADDITIONAL_CONNECTIONS;
+        }
+        this.totalRequestedConnections = this.localHostAddresses.size() + numAdditionalConnections;
+    }
+
+    @Override
+    public void init(Cluster cluster, Collection<Host> hosts)
+    {
+        this.cluster = cluster;
+        this.allHosts.addAll(hosts);
+        recalculateSelectedHosts();
+        childPolicy.init(cluster, hosts);
+    }
+
+    @Override
+    public HostDistance distance(Host host)
+    {
+        if (!selectedHosts.contains(host))
+        {
+            return HostDistance.IGNORED;
+        }
+        return childPolicy.distance(host);
+    }
+
+    @Override
+    public Iterator<Host> newQueryPlan(String loggedKeyspace, Statement statement)
+    {
+        Iterator<Host> child = childPolicy.newQueryPlan(loggedKeyspace, statement);
+        // Filter the child policy to only selected hosts
+        return Iterators.filter(child, selectedHosts::contains);
+    }
+
+    @Override
+    public synchronized void onAdd(Host host)
+    {
+        onUp(host);
+        childPolicy.onAdd(host);
+    }
+
+    @Override
+    public synchronized void onUp(Host host)
+    {
+        this.allHosts.add(host); // replace existing reference if there is one
+        if (selectedHosts.size() < totalRequestedConnections)
+        {
+            recalculateSelectedHosts();
+        }
+        childPolicy.onUp(host);
+    }
+
+    @Override
+    public synchronized void onDown(Host host)
+    {
+        // Don't remove local addresses from the selected host list
+        if (localHostAddresses.contains(host.getBroadcastRpcAddress()))
+        {
+            return;
+        }
```

Comment:
Local hosts should always be in the list even if they are down - the client will not use them if it thinks they are down, and this way we know they are always in the list and don't have to do anything more with down nodes.

I'm not sure if a node going down (local or otherwise) is worth a WARN - maybe debug, but it's not an uncommon occurrence so spitting out a WARN here seems overly noisy.